### ISSUE TITLE
twisted/conch/ssh/filetransfer.py misdefines FILEXFER_ATTR_EXTENDED

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3010]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3010|
|Type|defect|
|Created|2008-01-24 16:39:29Z|
|Branch|https://github.com/twisted/twisted/tree/filetransfer-attr-ext-3010|

Around line 862

```python
FILEXFER_ATTR_EXTENDED    = 0x+80000000
```

This will actually be a syntax error in Python 2.6, too.

<details><summary>Searchable metadata</summary>

```
trac-id__3010 3010
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__Python_2_6 Python-2.6
branch__branches_filetransfer_attr_ext_3010 branches/filetransfer-attr-ext-3010
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1201192769000000 1201192769000000
changetime__1205141345000000 1205141345000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set status to `closed`|
|-|-|

Duplicate of [#2519](https://github.com/twisted/twisted/issues/2519)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set status to `reopened`|
|-|-|

Actually, it's not.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [22709]) Branching to 'filetransfer-attr-ext-3010'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to z3p|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|z3p set owner to @therve|
|-|-|

Looks good to me.  Just update the copyright dates on those files and you're good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [22747]) Merge filetransfer-attr-ext-3010

Author: therve
Reviewer: exarkun
Fixes [#3010](https://github.com/twisted/twisted/issues/3010)

Fixe definition and management of extended attributes in conch file transfer.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `reopened`|
|-|-|

(In [22750]) Revert r22747: regression under Python 2.3

TestOurServerOurClient.test_openFileExtendedAttributes timeouts under Python
2.3, with failure:
exceptions.AssertionError: still have data in OPEN: '\x00\x00\x00\x01\x00\x00\x00\x03foo\x00\x00\x00\x03bar'

Reopens [#3010](https://github.com/twisted/twisted/issues/3010)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

This was fixed. Basically, it's because 0x80000000 != 0x80000000L under 2.3. Back to another turn of review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Why does `test_openFileExtendedAttributes` bother to call the base implementation?  Why doesn't it return the base implementation's result?  Why isn't that an error?  I guess since nothing is ever done with the file, it's okay that `None` is returned.  I'm torn between suggesting that the unneeded code should be deleted and suggesting that the sftp implementation should really be screaming loudly that `openFile` is returning `None`. :)  I lean a bit towards the former at the moment, since if we change the implementation later (including a test) then this code will start to fail in a very obvious way and be simple to fix.  Whereas if we do the latter then there'll just be some extra code that's not really doing anything useful yet.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [22757]) Don't call the base implementation.

Refs [#3010](https://github.com/twisted/twisted/issues/3010)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

Done!

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_13"></a>|@radix set owner to @therve|
|-|-|

So... I notice that if I change the value of the flag, nothing breaks. exarkun points out to me that there are already some tests which use a command line client against the Conch SFTP server; these may not be too hard to adapt to use the actual unix sftp command, to test our protocol support in this area for real. Do you think this is reasonable? If you don't want to implement this now, I guess it's OK, but at least another ticket should be filed. Other than that, the branch looks good, so please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [22782]) Add a functional test with sftp client, some cleanups.

Refs [#3010](https://github.com/twisted/twisted/issues/3010)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve removed owner|
|-|-|

OK, that should be good. Please review!

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_16"></a>|@radix set owner to @therve|
|-|-|

Thanks for adding this test, therve. When you merge [#3061](https://github.com/twisted/twisted/issues/3061) don't forget to remove the new use of setUpClass introduced in this branch :-)

+1!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve set status to `closed`|
|-|-|

(In [22785]) Merge filetransfer-attr-ext-3010

Author: therve
Reviewers: radix, z3p, exarkun
Fixes [#3010](https://github.com/twisted/twisted/issues/3010)

Fixe definition and management of extended attributes in conch file transfer.

This remerge fixes behavior with Python 2.3 and add a new test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set status to `reopened`|
|-|-|

(In [22790]) Revert r22785: Failure under FreeBSD

TestOurServerSftpClient.test_extendedAttributes fails with status code 255

Reopens [#3010](https://github.com/twisted/twisted/issues/3010)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

(In [22791]) Force host key algo.

Refs [#3010](https://github.com/twisted/twisted/issues/3010)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

We're gonna make it. I promise.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set owner to @therve<br>@glyph set status to `new`|
|-|-|

This change looks fine.  I checked on python2.3.  Merge.

(Pyflakes reveals a `NameError` on the (merged) line 58, but that's not introduced here.  Open another ticket, if there isn't one already?)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve set status to `closed`|
|-|-|

(In [22792]) Merge filetransfer-attr-ext-3010

Author: therve
Reviewers: glyph, radix, z3p, exarkun
Fixes [#3010](https://github.com/twisted/twisted/issues/3010)

Fixe definition and management of extended attributes in conch file transfer.

This re-remerge fixes behavior with FreeBSD.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
